#! /bin/sh -e

# All lines beginning with `# DP:' are a description of the patch.
# DP: Description: the Linux kernel always pads readdir entries, even on 32bit.
# DP: Bug: http://sourceware.org/bugzilla/show_bug.cgi?id=11333
# DP: Bug-Ubuntu: https://launchpad.net/bugs/392501
# DP: Origin: http://repo.or.cz/w/glibc.git/commitdiff/1a81139728494810f65aaa0d0c538ff8c2783dd5

PATCHLEVEL=1

if [ $# -ne 2 ]; then
    echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
    exit 1
fi
case "$1" in
    -patch) patch -d "$2" -f --no-backup-if-mismatch -p$PATCHLEVEL < $0;;
    -unpatch) patch -d "$2" -f --no-backup-if-mismatch -R -p$PATCHLEVEL < $0;;
    *)
	echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
	exit 1
esac
exit 0

Index: eglibc-2.10.1/sysdeps/unix/readdir_r.c
===================================================================
--- eglibc-2.10.1.orig/sysdeps/unix/readdir_r.c	2010-05-03 15:00:47.184052538 -0700
+++ eglibc-2.10.1/sysdeps/unix/readdir_r.c	2010-05-03 15:01:18.301553857 -0700
@@ -113,7 +113,20 @@
   while (dp->d_ino == 0);
 
   if (dp != NULL)
-    *result = memcpy (entry, dp, reclen);
+#ifndef MIN
+#define MIN(l,o) ((l) < (o) ? (l) : (o))
+#endif
+    {
+#ifdef GETDENTS_64BIT_ALIGNED
+      /* The d_reclen value might include padding which is not part of
+	 the DIRENT_TYPE data structure.  */
+      reclen = MIN (reclen, sizeof (DIRENT_TYPE));
+#endif
+      *result = memcpy (entry, dp, reclen);
+#ifdef GETDENTS_64BIT_ALIGNED
+      entry->d_reclen = reclen;
+#endif
+    }
   else
     *result = NULL;
 
Index: eglibc-2.10.1/sysdeps/unix/sysv/linux/i386/readdir64_r.c
===================================================================
--- eglibc-2.10.1.orig/sysdeps/unix/sysv/linux/i386/readdir64_r.c	2010-05-03 15:00:51.785303120 -0700
+++ eglibc-2.10.1/sysdeps/unix/sysv/linux/i386/readdir64_r.c	2010-05-03 15:01:18.301553857 -0700
@@ -19,6 +19,7 @@
 #define __READDIR_R __readdir64_r
 #define __GETDENTS __getdents64
 #define DIRENT_TYPE struct dirent64
+#define GETDENTS_64BIT_ALIGNED 1
 
 #include <sysdeps/unix/readdir_r.c>
 
